lib/engine/game/g_18_india/game.rb

```diff
@@ -656,13 +716,144 @@ def float_corporation(corporation)
         # Modified to allow yellow towns to be upgraded to SINGLE slot city green tiles
         # > Also modified legal_tile_rotation in STEP::Track or STEP::Tracker
         # Modified do prevent yellow cities upgrading to SINGLE slot city green tiles
+        # Allow GIPR Home Track directly to Green Single City
         def upgrades_to?(from, to, special = false, selected_company: nil)
+          return true unless @round.pending_tokens.empty?
           return true if yellow_town_to_city_upgrade?(from, to)
           return false if yellow_city_upgrade_is_single_slot_green?(from, to)
 
           super
         end
 
+        def gipr
+          @gipr ||= @corporations.find { |corp| corp.name == 'GIPR' }
+        end
+
+        def gipr_may_operate?
+          @phase.name != 'I' && gipr.floated?
+        end
+
+        def gipr_exchange_tokens
+          ability = gipr.all_abilities.find { |a| a.type == :exchange_token }
+          return 0 unless ability
+
+          ability.count
+        end
+
+        def gipr_has_exchange_token?
+          LOGGER.debug "gipr_may_exchange_token > #{gipr.floated?} && #{gipr_exchange_tokens.positive?}"
+          gipr.floated? && gipr_exchange_tokens.positive?
+        end
+
+        def use_gipr_exchange_token
+          ability = gipr.all_abilities.find { |a| a.type == :exchange_token }
+          ability.use!
+          ability.description = "Exchange tokens: #{ability.count}"
+        end
+
+        def gipr_exchange_with_closing_corp(corporation)
+          corporation.tokens.dup.each_with_index do |token, index|
+            LOGGER.debug "gipr_exchange > #{corporation.tokens} && index-#{index} used-#{token.used}"
+            break unless token.used
```

Comment:
I'll see if I can use placed_tokens, as that should remove the need for the `used` guard clause.   I think I needed the `dup` because I was deleting the tokens as I went through the list.   I recall this flow being a bit complex as `close_corporation` would get called again with the way pending tokens worked, so I needed a way to handle re-entering that method to check if we were already did the exchange token test.

Reply:
Using` corporation.placed_tokens` works.  I also removed the `dup` call and changed the token `destroy` to `remove`, which just modifies the token.  Seems to work find in testing.